# Section 7 PP240A/DEVP209

In this notebook, we will review some of the tools we learned in past weeks for describing and plotting data.

## Import modules

As usual, we import the modules we will want to call later at the top of our file. If you don't run the cell below, then the rest of the code won't work. Remember that we have to import modules first before we can make use of functions contained in them.

In [2]:
import pandas as pd                 # module for manipulating DataFrames
import datetime                     # module for manipulating dates and times
import numpy as np                  # module for manipulating arrays  
import matplotlib.pyplot as plt     # module for creating graphs
import seaborn as sns               # module for statistical data visualization
import statsmodels.api as sm        # module for statistical models
from scipy.stats import norm        # module for finding cumulative densities of data

## Review

Run and make sure you understand the code cells in this section.

Just run the cells in this section (no need to write anything). Please make sure you understand what each cell is doing.

In [3]:
# Load the data
removals = pd.read_csv('removals.csv')

# Display the first few rows of the DataFrame
removals.head()

,Departed Date,Port of Departure,Departure Country,Gender,Citizenship Country,Birth Year,Entry Date,Entry Status,Felon,Final Program,Prior Deport Yes No,Latest Person Departed Date
0,07/04/2024,"NOGALES, AZ, POE",MEXICO,Male,MEXICO,1991,06/05/2024,PWA Mexico,NaN,Border Patrol,YES,07/04/2024
1,09/09/2023,"CALEXICO, CA, POE",MEXICO,Male,MEXICO,1995,06/15/2023,PWA Mexico,NaN,Border Patrol,YES,09/09/2023
2,09/09/2023,"NOGALES, AZ, POE",MEXICO,Female,MEXICO,1993,08/22/2023,PWA Mexico,NaN,Border Patrol,YES,09/09/2023
3,06/25/2024,"HARLINGEN, TX, POE",COLOMBIA,Male,COLOMBIA,2000,06/09/2024,PWA Mexico,NaN,Border Patrol,NO,06/25/2024
4,04/02/2025,"NOGALES, AZ, POE",MEXICO,Male,MEXICO,1980,11/27/2024,Other Applicant for Admission,Not an Aggravated Felon,Inspections - Land,NO,04/02/2025


In [4]:
# Reminder 1: selecting a column from a DataFrame

# Recall that we select a column using the following 
departureCountryColumn = removals['Departure Country']

# If we want to know all of the countries that people are being removed to, we can use the following
uniqueDepartureCountries = removals['Departure Country'].unique()
print(uniqueDepartureCountries)


['MEXICO' 'COLOMBIA' 'GUYANA' 'ECUADOR' 'EL SALVADOR' 'NICARAGUA' 'ANGOLA'
 'HONDURAS' 'SENEGAL' 'GUATEMALA' 'BRAZIL' 'UNITED KINGDOM'
 'CHINA, PEOPLES REPUBLIC OF' 'VENEZUELA' 'DOMINICAN REPUBLIC' 'ETHIOPIA'
 'PERU' 'CHILE' 'HAITI' 'PAKISTAN' 'BARBADOS' 'NEPAL' 'CANADA' 'MOROCCO'
 'PHILIPPINES' 'PANAMA' 'COSTA RICA' 'GHANA' 'GEORGIA' 'ROMANIA'
 'ARGENTINA' 'EGYPT' 'TURKIYE' 'RUSSIA' 'ARMENIA' 'JORDAN' 'BURKINA FASO'
 'IVORY COAST' 'BOLIVIA' 'MAURITANIA' 'VIETNAM' 'ALBANIA' 'JAMAICA' 'CUBA'
 'INDIA' 'BANGLADESH' 'GERMANY' 'IRAN' 'NIGERIA' 'POLAND' 'IRELAND'
 'GUINEA' 'UZBEKISTAN' 'BOSNIA-HERZEGOVINA' 'BELIZE' 'SPAIN' 'DENMARK'
 'SAUDI ARABIA' 'SYRIA' 'SABA' 'KUWAIT' 'ZAMBIA' 'BAHAMAS'
 'CZECH REPUBLIC' 'BURUNDI' 'QATAR' 'TAJIKISTAN' 'BELARUS' 'KYRGYZSTAN'
 'SIERRA LEONE' 'MICRONESIA, FEDERATED STATES OF' 'CAMBODIA' 'IRAQ'
 'BHUTAN' 'ZIMBABWE' 'MOLDOVA' 'HOLY SEE' 'PORTUGAL' 'TAIWAN'
 'TRINIDAD AND TOBAGO' 'NETHERLANDS' 'KOSOVO' 'BURMA' 'GUADELOUPE' 'CHAD'
 'CAMEROON' 'INDONESIA' 'ITALY

In [5]:
# There are a lot of countries listed, too many to be read in the output. Let us see how many there are:
print(len(uniqueDepartureCountries))

# Notice that this is also the way we find the total number of rows in the dataframe
print(len(removals))

211
483263


In [6]:
# Reminder 2: filtering rows in a DataFrame

# We want to learn what is in the data frame. 
mexico_removals = removals.loc[removals['Citizenship Country'] == 'MEXICO']

# Now we can look at the subset data; notice that the index goes 0,1,2,4,8 because it is skipping the rows that were not MEXICO
mexico_removals.head()


,Departed Date,Port of Departure,Departure Country,Gender,Citizenship Country,Birth Year,Entry Date,Entry Status,Felon,Final Program,Prior Deport Yes No,Latest Person Departed Date
0,07/04/2024,"NOGALES, AZ, POE",MEXICO,Male,MEXICO,1991,06/05/2024,PWA Mexico,NaN,Border Patrol,YES,07/04/2024
1,09/09/2023,"CALEXICO, CA, POE",MEXICO,Male,MEXICO,1995,06/15/2023,PWA Mexico,NaN,Border Patrol,YES,09/09/2023
2,09/09/2023,"NOGALES, AZ, POE",MEXICO,Female,MEXICO,1993,08/22/2023,PWA Mexico,NaN,Border Patrol,YES,09/09/2023
4,04/02/2025,"NOGALES, AZ, POE",MEXICO,Male,MEXICO,1980,11/27/2024,Other Applicant for Admission,Not an Aggravated Felon,Inspections - Land,NO,04/02/2025
8,03/21/2024,"EL PASO, TX, POE",MEXICO,Male,MEXICO,2001,03/12/2024,PWA Mexico,NaN,Border Patrol,NO,03/21/2024


In [7]:
# Reminder 3: grouping data in a DataFrame

# Let's group the dataframe by apprehensionstate and count the number of arrests in each state
removals_by_citizenship_country = removals.groupby('Citizenship Country').size().reset_index(name='counts')

# let's break down each part of the above 
# arrests_state_country.groupby('apprehensionstate') groups the dataframe by the values in the 'apprehensionstate' column
# .size() counts the number of rows in each group
# .reset_index(name='counts') resets the index of the resulting dataframe and names the counts column 'counts'

# print the head of arrests_by_state to see the result
print(removals_by_citizenship_country.head())

  Citizenship Country  counts
0         AFGHANISTAN     344
1             ALBANIA      92
2             ALGERIA      32
3              ANGOLA     273
4            ANGUILLA       1


In [8]:
# Reminder 4: finding conditional probability from a data frame
conditional_probs_of_removal = removals.groupby('Gender')['Departure Country'].value_counts(normalize=True).reset_index(name='conditional probability')

# let's break down each part of the above 
# removals.groupby('Departure Country') groups the dataframe by the values in the 'Gender' column; all 'Female' rows are considered together
# ['Departure Country'] does a column selection, showing that for each gender ['Male','Female','Unknown'], we find all of the countries represented for each
# .value_counts(normalize=True) finds the count of each 'Departure Country' for that gender, than divides by the total number of rows in the removals data frame for that gender
# .reset_index(name='conditional probability') resets the index of the resulting dataframe and names the counts column 'conditional probability'

conditional_probs_of_removal


,Gender,Departure Country,conditional probability
0,Female,GUATEMALA,0.267647
1,Female,MEXICO,0.211028
2,Female,HONDURAS,0.197159
3,Female,EL SALVADOR,0.076368
4,Female,COLOMBIA,0.069221
...,...,...,...
395,Unknown,PERU,0.008197
396,Unknown,POLAND,0.008197
397,Unknown,SOUTH KOREA,0.008197
398,Unknown,TRINIDAD AND TOBAGO,0.008197


### Warm-up questions

Solutions in a markdown cell below

In [9]:
arrests_state_country = pd.read_csv('ArrestsStateCountry.csv', sep='\t')
arrests_state_country.head()

,apprehensionstate,citizenshipcountry,arrestsbiden,arrests_perdaybiden,arreststrump,arrests_perdaytrump,popACS2023,popACS2024,ln2023,ln2024,predicted,predicted_level,pop_combined,pct_peryear_biden,pct_peryear_trump
0,ALABAMA,BRAZIL,1.0,0.001976,2.0,0.010526,1646.0,NaN,7.406104,NaN,7.751637,2325.376,2325.376,0.000310,0.001652
1,ALABAMA,CHINA,10.0,0.019763,3.0,0.015790,10355.0,NaN,9.245225,NaN,9.263822,10549.370,10549.370,0.000684,0.000546
2,ALABAMA,COLOMBIA,33.0,0.065217,19.0,0.100000,2949.0,NaN,7.989222,NaN,8.436029,4610.213,4610.213,0.005163,0.007917
3,ALABAMA,CUBA,14.0,0.027668,15.0,0.078947,2740.0,NaN,7.915713,NaN,8.358318,4265.516,4265.516,0.002367,0.006756
4,ALABAMA,DOMINICAN REPUBLIC,11.0,0.021739,7.0,0.036842,1701.0,NaN,7.438972,NaN,7.441265,1704.906,1704.906,0.004654,0.007888


In [10]:
# Q1: how many rows correspond to 'Final Program' of 'Border Patrol'?

In [12]:
# Q2: What is the conditional probability of having a 'Final Program' NOT of 'Border Patrol' if the gender is 'Female'?

In [15]:
# Q3: generate a new dataframe with rows where the 'Final Program' is 'Border Patrol' and the gender is 'Male'
# how many rows are there in this new dataframe? How many unique 'Departure Country' values are represented? Is this fewer countries than all included?

### Solutions

Q1: 339101

Q2: 0.10398207660023114

Q3: 250680 rows long; 172 countries represented compared to the 211 countries in the original removals DataFrame

## Using Python for confidence intervals

Let us practice using Python to calculate p-values and confidence intervals without using the complicated tables

Above, we imported from scipy.stats the norm library, which allows us to calculate values related to the *Normal Distribution*

Quick warm up:

Q1: What does the cdf stand for? What does it represent?

Q2: What is a confidence interval? Write down in your own words and the equation for a generic confidence interval (meaning not necessarily 90% or 95%)

Answers:

Q1: Cumulative density function. When we find the cumulative density, it tells us the total percentage of a distribution that occurs below a specific value. We are also able to express a range in the CDF function that tells us the total area under the probability curve in between two values (such as a confidence interval). A famous example is the total area below -1.96 and above 1.96 (WHY???? see below)

Q2: A confidence interval is the range that we expect a true value to lie within given a certain probability. If I have a 95% confidence interval, I expect a value that I measure for the mean to occur within a LOWER and UPPER BOUND to occur 95% of the time. We can still get outliers that are outside of this range

In [17]:
# Please think of the answers before running this cell using what you know about confidence intervals

# Example 1: P(Z <= 1.96) #remember this is a one-sided confidence interval
p = norm.cdf(1.96)
print("P(Z <= 1.96) =", p)

# Example 2: P(Z > 1.96)
print("P(Z > 1.96) =", 1 - p)

# Example 3: z critical value for a 95% confidence level
z_crit = norm.ppf(0.975)
print("z critical value (95% CI) =", z_crit)

P(Z <= 1.96) = 0.9750021048517795
P(Z > 1.96) = 0.024997895148220484
z critical value (95% CI) = 1.959963984540054


Okay so what we see above, as we remember from class is that a 95% confidence interval has a critical value of 1.96 (example 3)

A 95% confidence interval has an UPPER and LOWER BOUND, so the area outside of each of these bounds is 5%/2 = 2.5%, which is why Example 2 is 2.5%

Lastly, Example 1 has a value of 97.5% because we are not looking at the lower bound here of -1.96, so the value is 95% plus the 2.5% extra not included

Exercises:

Q1: Compute P(−1.96<𝑍<1.96) using the formulas from above using `norm.cdf`
    

Q2: Change 1.96 to 2.58. What happens to the probability?

Q3: What is the critical z-value for a 99% confidence interval? (remember, the remaining 1% is split across the upper and lower side; also use `norm.ppf`)

Answers:

### Plotting the normal distribution

In [ ]:
x = np.linspace(-4, 4, 400)
y = norm.pdf(x)

plt.plot(x, y, label='Standard Normal PDF')
plt.fill_between(x, 0, y, where=(x > -1.96) & (x < 1.96), color='skyblue', alpha=0.5, label='95% Area')
plt.title("Standard Normal Distribution")
plt.xlabel("Z")
plt.ylabel("Density")
plt.legend()
plt.show()

Q1: Change -1.96 and 1.96 to represent a 90% confidence interval instead.

Q2: What happens to the shaded area in your own words?

Shaded area gets smaller, meaning lower probability covered

### Challenge:

Change -1.96 and 1.96 above to represent a 45% confidence interval (you will probably need to use `norm.ppf`)

## Sample Mean

In [ ]:
# Suppose we have a population with mean 50 and standard deviation 10.
# We draw a random sample of n = 100.

population_mean = 50
population_sd = 10
n = 100

sample = np.random.normal(loc=population_mean, scale=population_sd, size=n) #We make a random normal distribution around our population data
sample_mean = np.mean(sample)
sample_var = np.var(sample, ddof=2)

print("Sample mean:", sample_mean)
print("Sample variance:", sample_var)


Note, in the above, ddof is what is subtracted from the divisor n in the variance formula. If you remember, for a sample, we divide by n-1

Q1: Run this cell multiple times — does the sample mean always equal 50?

Q2: Increase n to 1000 — what happens to how close your mean is to 50?

## Law of Large Numbers

In [ ]:
# Let's simulate many sample means to see how they behave
num_samples = 1000
n = 30
sample_means = []

for i in range(num_samples):
    x = np.random.normal(population_mean, population_sd, n)
    sample_means.append(np.mean(x))

plt.hist(sample_means, bins=30, density=True, alpha=0.7)
plt.axvline(population_mean, color='red', linestyle='dashed', label='True Mean')
plt.title("Sampling Distribution of the Sample Mean (n=30)")
plt.xlabel("Sample Mean")
plt.ylabel("Density")
plt.legend()
plt.show()

print("Mean of sample means:", np.mean(sample_means))
print("SD of sample means:", np.std(sample_means))


Q1: Change n to 10, then to 100. What happens to the spread of the histogram?

Q2: What does this show about the Law of Large Numbers?

## Word problem question

In [ ]:
# A researcher collects wages (in $/hour) for 40 workers
xbar = 22.5
sigma = 4.5
n = 40

z_crit = norm.ppf(0.95)  # for 90% CI
margin = z_crit * sigma / np.sqrt(n)
ci = (xbar - margin, xbar + margin)

print("90% Confidence Interval for average wage:", ci)


Q1: Find the 88% confidence interval and interpret in plain language

## Sample Variance

In [ ]:
from scipy.stats import t

# Example: Small sample, unknown sigma
x = np.array([45, 50, 47, 52, 44, 49, 46, 51])  # sample data
n = len(x)
xbar = np.mean(x)
s = np.std(x, ddof=1)  # sample standard deviation

confidence = 0.95
alpha = 1 - confidence

t_crit = t.ppf(1 - alpha/2, df=n-1)
margin = t_crit * s / np.sqrt(n)
ci = (xbar - margin, xbar + margin)

print("Sample mean:", round(xbar, 2))
print("Sample standard deviation:", round(s, 2))
print(f"{int(confidence*100)}% CI for mean (using t):", ci)

Q1: Add more data points to the sample x. (write whatever numbers you want, but try to keep them centered aroudn 50). How does the width of the confidence interval change as the n increases?

Q2: Replace `t.ppf` with `norm.ppf` (you will need to remove df=n-1). What do these functions do? What happens to the interval width and why?

Q3: Explain in words why the t-interval is wider for smaller samples.

## Bonus: Why use Python?

So a lot of the above, you could have just done with a pencil, piece of paper, and normal distribution look up table. So why is Python important?

Here are some examples of why Python is valuable for data analysis, doing things that we otherwise couldn't

Example 1: 

Suppose test scores are normally distributed with
mean μ = 72 and standard deviation σ = 9.

What is the probability that a randomly chosen student scores above 85?

Method 1:
With paper, we could see Z = (X-mu)/sigma = (85-72)/9 = 1.44

P(X>85) = P(Z>1.44) = 1- 0.9251 (found in table) = 0.0749 meaning ~7.5% of students performed above 85 on this test

In [30]:
# Define variables for Python method
mu = 72
sigma = 9
x_value = 85

In [31]:
# Method 2: Use python and get answer immediately

# Use norm.cdf to get cumulative probability up to 85
p_below = norm.cdf(x_value, loc=mu, scale=sigma)
p_above = 1 - p_below

print("P(X > 85) =", p_above)

P(X > 85) = 0.07430699822760589


In [32]:
# Method 3: Use python to convert to Z and find answer more slowly

z = (x_value - mu) / sigma
p_above_z = 1 - norm.cdf(z)
print("P(X > 85) using standard normal:", p_above_z)


P(X > 85) using standard normal: 0.07430699822760589


### Example 2: What is the probability that a student's score lies between 60 and 90

In [33]:
p_between = norm.cdf(90, mu, sigma) - norm.cdf(60, mu, sigma)
print("P(60 < X < 90) =", p_between)

P(60 < X < 90) = 0.8860386483259529


### Questions:
Q1: Find P(X<65)

Q2: Find P(X>95)

## Congrats

You are done with this section. What we learned:
* How to use python to quickly do the calculations that are very hard with calculator for distributions
* Central Limit Theorem and Law of Large Numbers
* Student's t for unknown variance and normal distribution

What to do now? Start working on your group project. Find a data set and talk to your team